R/preprocess.R

```diff
@@ -21,41 +21,142 @@ enw_metadata <- function(obs, target_date = "reference_date") {
+#'
+#' However, it can also be used directly on other data.
+#'
+#' @return A copy of the `metaobs` input, with additional columns:
+#'  * `day_of_week`, a factor of values as output from [weekdays()] and
+#'  possibly as `holiday_to` if distinct from weekdays values
+#'  * `day`, numeric, 0 based from start of time series
+#'  * `week`, numeric, year based
+#'  * `month`, numeric, year based
+#'
 #' @family preprocess
+#' @importFrom purrr compose
 #' @export
-#' @importFrom data.table as.data.table
-enw_add_metaobs_features <- function(metaobs, holidays = c(),
-                                     holidays_to = "Sunday") {
-  # add days of week
-  metaobs <- data.table::copy(metaobs)
-  metaobs[, day_of_week := weekdays(date)]
-
-  # make holidays be Sundays
-  if (length(holidays) != 0) {
-    metaobs[get(holidays) == TRUE, day_of_week := holidays_to]
+#' @examples
+#'
+#' # make some example date
+#' nat_germany_hosp <- subset(
+#'   germany_covid19_hosp,
+#'   location == "DE" & age_group == "80+"
+#' )[1:40]
+#'
+#' basemeta <- enw_add_metaobs_features(
+#'   nat_germany_hosp,
+#'   datecol = "report_date"
+#' )
+#' basemeta
+#'
+#' # with holidays - n.b.: holidays not found are silently ignored
+#' holidaymeta <- enw_add_metaobs_features(
+#'   nat_germany_hosp,
+#'   datecol = "report_date",
+#'   holidays = c(
+#'     "2021-04-04", "2021-04-05",
+#'     "2021-05-01", "2021-05-13",
+#'     "2021-05-24"
+#'   ),
+#'   holidays_to = "Holiday"
+#' )
+#' holidaymeta
+#' subset(holidaymeta, day_of_week == "Holiday")
+enw_add_metaobs_features <- function(
+  metaobs,
+  holidays = c(),
+  holidays_to = "Sunday",
+  datecol = "date"
+) {
+  # localize and check metaobs input
+  metaobs <- data.table::as.data.table(metaobs)
+  if (is.null(metaobs[[datecol]])) {
+    stop(sprintf("metaobs does not have datecol '%s'.", datecol))
+  } else if (!is.Date(metaobs[[datecol]])) {
+    stop(sprintf("metaobs column '%s' is not a Date.", datecol))
   }
 
-  # make day of week a factor
-  metaobs[, day_of_week := factor(day_of_week)]
+  # this may also error, so coercing first
+  holidays <- coerce_date(holidays)
 
-  # add day feature
-  metaobs[, day := as.numeric(date)]
-  metaobs[, day := day - min(day)]
+  # warn about columns that may be overwritten
+  tarcols <- c("day_of_week", "day", "week", "month")
+  if (any(tarcols %in% colnames(metaobs))) {
+    warning(sprintf(
+      "Pre-existing columns in `metaobs` will be overwritten: {%s}.",
+      intersect(tarcols, colnames(metaobs))
+    ))
+  }
 
-  # add week feature
-  metaobs[, week := lubridate::week(date)]
-  metaobs[, week := week - min(week)]
+  if (is.unsorted(metaobs[[datecol]])) {
+    warning(sprintf(
+      "metaobs is not sorted by column '%s'; returned result will be.",
+      datecol
+    ))
+  }
+  setkeyv(metaobs, union(key(metaobs), datecol))
+
+  # function to transform numbers to be referenced from 0
+  zerobase <- ~ .x - min(.x)
```

Comment:
can these be proper internal functions. I think it's fine to have the code be a few lines longer.